### ISSUE TITLE
Missing screen reader output at opening / saving file input on remote

### ISSUE THREAD

**Issue by yncat** [microsoft/vscode#237324]
Type: <b>Bug</b>

If there is a workaround, it may not necessarily be a bug.

1. Open a remote using the vscode-remote extension. E.G. WSL.
2. Try to open a folder, or save a new file.
3. Try to input filename, or edit the current input with screen reader.
4. inputs and deletions are not announced by screen reader at all. Moving cursor doesn't announce the focused character.
5. Expected: They are announced.

I think this box was originally a simple text field. In a older version (probably 6 month to 1 year ago?), it also started to act as a combo box, allowing me to autocomplete directories.
After the change, I am unable to smoothly finalize the destination to what I want the path to be. I just thought I would be able to adapt to this when it was introduced, but now feeling like reporting it here and seek for some workarounds.

It does not occur in local development since it uses the native dialog, at least on Windows. It only happens when I am developing on a remote.

I'm not really sure this custom dialog is provided by the vscode-side or the extension-side.

Extension version: 0.88.5
VS Code version: Code 1.96.1 (42b266171e51a016313f47d0c48aca9295b9cbb2, 2024-12-17T17:50:05.206Z)
OS version: Windows_NT x64 10.0.22631
Modes:
Remote OS version: Linux x64 5.15.167.4-microsoft-standard-WSL2

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|12th Gen Intel(R) Core(TM) i7-1255U (12 x 2611)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled<br>webnn: disabled_off|
|Load (avg)|undefined|
|Memory (System)|15.69GB (2.62GB free)|
|Process Argv|--crash-reporter-id 16c020cb-6298-45d1-bf2f-54354940af72|
|Screen Reader|yes|
|VM|0%|

|Item|Value|
|---|---|
|Remote|WSL: Ubuntu|
|OS|Linux x64 5.15.167.4-microsoft-standard-WSL2|
|CPUs|12th Gen Intel(R) Core(TM) i7-1255U (12 x 0)|
|Memory (System)|7.60GB (4.88GB free)|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vspor879:30202332
vspor708:30202333
vspor363:30204092
vswsl492:30256859
vscod805:30301674
binariesv615:30325510
vsaa593cf:30376535
py29gd2263:31024239
c4g48928:30535728
azure-dev_surveyone:30548225
962ge761:30959799
pythonnoceb:30805159
pythonmypyd1:30879173
h48ei257:31000450
pythontbext0:30879054
cppperfnew:31000557
dsvsc020:30976470
pythonait:31006305
dsvsc021:30996838
dvdeprecation:31068756
dwnewjupytercf:31046870
nativerepl2:31139839
pythonrstrctxt:31112756
nativeloc2:31192216
cf971741:31144450
iacca1:31171482
notype1cf:31157160
5fd0e150:31155592
dwcopilot:31170013
stablechunks:31184530
6074i472:31201624

```

</details>

**Comment by yncat** [comment]
Thanks for triaging the issue. Just forgot to report that I was using NVDA to reproduce this issue. Just FYI!

**Comment by TylerLeonhardt** [comment]
@alexr00 No major changes in quick pick for over a year now.

btw @yncat you can force the simple file picker when not on a remote by applying this setting:
```
"files.simpleDialog.enable": true
```

Though, this ultimately could have been from a change way back when... 2 years? We changed things so that screen reader users didn't get a different experience to non-screen reader users.

Previously, screen reader users didn't have the first item in a quick pick "active". This was to help with reading out what was in the input. We went back and forth with some screen reader users and ultimately ended on the current solution which was deemed the best option:
* on-type, have the first item be active (this is the non-screen reader experience)
* allow the user to hit Shift-Tab which will clear the active item allowing a screenreader to read what's in the input.

**Comment by yncat** [comment]
@TylerLeonhardt 
So, it is assumed that I start typing there and press shift + tab to review or edit what I have just typed.
Based on the expected behavior above, here is what is happening for me:

1. Try to "save as" a new file.
2. Start typing
3. press shift+tab

Expected: Clears the active selection and lets me edit the input.
Happening: Nothing happens, looks like the selection is still active, no screenreader / focus updates.

As you wrote, the SimpleDialog option doesn't work on remote, so I'm still having a bit of difficulty especially when saving a new file. It works just fine locally.

**Comment by TylerLeonhardt** [comment]
Here's what I'm seeing using NVDA:

1. Using WSL to get a Remote-like experience
1. "save as" an Untitled file to show Simple File Dialog
1. I get the file path in the editor along with `Untitled-1` highlighted
1. 🐛 Bug 1: It says "21 results" but not the title of the SimpleDialog "Save as"
1. As I type, `asdf` it reads the characters
1. 🐛 Bug 2: backspace does not read out anything
2. 🐛 Bug 3: arrow keys moving to the left and right of the input are not read out

During this entire test, there was no item in the SimpleFileDialog that was selected (aka no item was highlighted) so we are already in what I'll refer to as "input focus mode" where Bug 2 and Bug 3 _should_ work. Something must have changed at some point around our input boxes.

@meganrogge anything jogging you memory? This use to work.

As for Bug 1... we do set the aria label of the input box to include the title, but I'm thinking _maybe_ the fact that there's input already in the box is maybe showing off Bug 1.

**Comment by meganrogge** [comment]
@TylerLeonhardt does this reproduce with edit context disabled? 

That's the only thing that has changed recently AFAIK that would cause issues for the screen reader reading the text area.

**Comment by TylerLeonhardt** [comment]
No. The input box in the quick pick is not a Monaco editor, just a normal input field.

**Comment by TylerLeonhardt** [comment]
Ok, I got to the bottom of this... I swear this use to all work. My only suspicion is that NVDA changed how they handle `role="combobox"... which is what we currently have the `input` element set to in the quick pick.

If I change this to `textbox` or `searchbox`, the good behavior I described starts happening. What makes this role different? I really don't know for sure.

In any case, I merged in the change to `textbox` since `searchbox` was added in ARIA 1.2... and I'd rather maximize for compatibility.

If you or anyone knows why this behavior is the way it is, let me know! Or, if you have an idea of how I can get the screen reader to announce the changed in the input, I'd loveeeee to know.

**Comment by meganrogge** [comment]
@rperez030 any insights?

**Comment by rperez030** [comment]
> [@rperez030](https://github.com/rperez030) any insights?

An element with the role of "combobox" is not intended to accept direct text input. This is why an editable combobox requires both a combobox element to host the options and a separate textbox for value editing. Additionally, if I recall correctly, this component relies on `aria-activedescendant`, which overrides keyboard focus when updated. In other words, the screen reader will not track changes in the caret position if `aria-activedescendant` indicates that focus is on one of the options. 

In this particular component, pressing Shift + Tab from the list of options is supposed to move focus to the editable portion of the widget for manual editing. However, since this portion uses the role "combobox," it does not function as expected.

[This example demonstrates the use of a text box with an element with role combobox to create an editable combobox](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-autocomplete-list/)

**Comment by TylerLeonhardt** [comment]
Interesting idea, @rperez030... For left & right, that does make a lot of sense... we could remove `aria-activedescendant` for this. If we do this, how would you expect we re-add `aria-activedescendant`? Perhaps when the user presses up or down arrow?

For delete & backspace... these will cause the filter to be different and as a result, it will cause the first item in the list to be focused and `aria-activedescendant` will be added... so I'm not sure about these. Maybe actually this is ok because the screen reader will read the deletion and then read the newly focused item. 🤔 

Holistically, I wondered if we could remove `aria-activedescendant`, allow the input to say what it needs to say, and then re-add `aria-activedescendant`... so that we don't have to have any Shift+Tab like behavior and the behavior is "the same" for screen reader users and not... but I'm not sure what the work involved would be.

**Comment by rperez030** [comment]
Left or right arrow keys, as well as any of the deletion keys (Delete or Backspace), could work exactly as Shift + Tab works right now. Typing anything or pressing the Up or Down arrow key would re-add `aria-activedescendant` to the item highlighted in the list, exactly as it does now. I believe removing the use of `aria-activedescendant` entirely would reduce the efficiency of using the component with a screen reader since this is the way to quickly determine which item is highlighted—unless you manage keyboard focus directly.

In summary, the only change that may be needed is to add support for Backspace, Delete, and Left and Right arrow keys to behave like Shift + Tab.

**Comment by TylerLeonhardt** [comment]
@rperez030 I have implemented something like this in https://github.com/microsoft/vscode/pull/239857

**Comment by rperez030** [comment]
@yncat I would love to hear your thoughts on the new behavior in insiders.

**Comment by yncat** [comment]
Thank you very much everyone for the discussion and implementation! I'm definitely going to test it in the insiders, when I work on coding next time. Tonight or tomorrow!

**Comment by yncat** [comment]
I've just finished the testing, and it's working perfectly for me!

Comparison: 1.97.0 and 1.98.0-insiders `fe4d6a831a4f87126187ac6ba96e16405d69de0f`

- Pressing down and up arrows still allows me to choose from suggestions
- Detects left arrow, right arrow, backspace or character inputs and quickly changes the context so that it can announce the information I need
- During the input mode, pressing down or up arrow switches back to the combobox-like behavior, probably reattaching the proper aria roles
- The shift + tab command doesn't work as originally explained in this thread, but it is no longer needed IMO, since combining arrow keys and inputting makes every action in this screen accessible as I understand

**Comment by TylerLeonhardt** [comment]
Yes, I removed the shift+tab behavior since it didn't feel needed anymore. Which is great as removing it closed out another issue I had 😁
